diff --git a/core/iwasm/libraries/lib-socket/inc/wasi_socket_ext.h b/core/iwasm/libraries/lib-socket/inc/wasi_socket_ext.h
index 98429bbb..488edb0e 100644
--- a/core/iwasm/libraries/lib-socket/inc/wasi_socket_ext.h
+++ b/core/iwasm/libraries/lib-socket/inc/wasi_socket_ext.h
@@ -135,6 +135,18 @@ typedef struct __wasi_addr_info_hints_t {
 #define IPV6_LEAVE_GROUP 21
 #define IPV6_V6ONLY 26
 
+struct hostent
+{
+    char *h_name;             /* Official name of host.  */
+    char **h_aliases;         /* Alias list.  */
+    int h_addrtype;           /* Host address type.  */
+    int h_length;             /* Length of address.  */
+    char **h_addr_list;       /* List of addresses from name server.  */
+#ifdef __USE_MISC
+#define h_addr h_addr_list[0] /* Address, for backward compatibility.*/
+#endif
+};
+
 struct addrinfo {
     int ai_flags;             /* Input flags.  */
     int ai_family;            /* Protocol family for socket.  */
